src/devices/Arduino/FirmataDevice.cs

```diff
@@ -0,0 +1,1234 @@
+                                    return;
+                                }
+
+                                int idx = 1;
+                                int pin = 0;
+                                while (idx < raw_data.Length)
+                                {
+                                    if (raw_data[idx] != 127)
+                                    {
+                                        _supportedPinConfigurations[pin].AnalogPinNumber = raw_data[idx];
+                                    }
+
+                                    idx++;
+                                    pin++;
+                                }
+
+                                _dataReceived.Set();
+                            }
+
+                            break;
+                        case FirmataSysexCommand.I2C_REPLY:
+
+                            _lastResponse = raw_data;
+                            _dataReceived.Set();
+                            break;
+
+                        case FirmataSysexCommand.SPI_DATA:
+                            _lastResponse = raw_data;
+                            _dataReceived.Set();
+                            break;
+
+                        case FirmataSysexCommand.DHT_SENSOR_DATA_REQUEST:
+                        case FirmataSysexCommand.PIN_STATE_RESPONSE:
+                            _lastResponse = raw_data; // the instance is constant, so we can just remember the pointer
+                            _dataReceived.Set();
+                            break;
+
+                        default:
+
+                            // we pass the data forward as-is for any other type of sysex command
+                            break;
+                    }
+
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Replaces the first occurrence of search in input with replace.
+        /// </summary>
+        private static string ReplaceFirst(String input, string search, string replace)
+        {
+            int idx = input.IndexOf(search, StringComparison.InvariantCulture);
+            string output = input.Remove(idx, search.Length);
+            output = output.Insert(idx, replace);
+            return output;
+        }
+
+        /// <summary>
+        /// Simulates a printf C statement.
+        /// Note that the word size on the arduino is 16 bits, so any argument not specifying an l prefix is considered to
+        /// be 16 bits only.
+        /// </summary>
+        /// <param name="fmt">Format string (with %d, %x, etc)</param>
+        /// <param name="bytesReceived">Total bytes received</param>
+        /// <param name="startOfArguments">Start of arguments (first byte of formatting parameters)</param>
+        /// <returns>A formatted string</returns>
+        private string PrintfFromByteStream(string fmt, in Span<byte> bytesReceived, int startOfArguments)
+        {
+            string output = fmt;
+            while (output.Contains("%"))
```

Comment:
you might benefit from using StreamReader here

Reply:
Doesn't work, at least not out of the box (I'm kind of "in-place" updating the string). Should probably be using StringBuilder, but several methods are missing there (i.e. there's no StringBuilder.Contains) and converting back and forth doesn't really help then as well.